src/RequestHandler/outofprocess/serverprocess.cxx

```diff
@@ -1123,36 +1134,36 @@ SERVER_PROCESS::CheckIfServerIsUp(
 
     if (!g_fNsiApiNotSupported)
     {
-        dwResult = GetExtendedTcpTable(NULL,
-                                       &dwSize,
-                                       FALSE,
-                                       AF_INET,
-                                       TCP_TABLE_OWNER_PID_LISTENER,
-                                       0);
-
-        if (dwResult != NO_ERROR && dwResult != ERROR_INSUFFICIENT_BUFFER)
+        while (dwResult == ERROR_INSUFFICIENT_BUFFER)
         {
-            hr = HRESULT_FROM_WIN32(dwResult);
-            goto Finished;
-        }
+            // Increase the buffer size with additional space, MIB_TCPROW 20 bytes
+            // New entries may be added by other processes before calling GetExtendedTcpTable
+            dwSize += 200;
 
-        pTCPInfo = (MIB_TCPTABLE_OWNER_PID*)HeapAlloc(GetProcessHeap(), 0, dwSize);
-        if (pTCPInfo == NULL)
-        {
-            hr = E_OUTOFMEMORY;
-            goto Finished;
-        }
+            if (pTCPInfo != NULL)
+            {
+                HeapFree(GetProcessHeap(), 0, pTCPInfo);
+            }
 
-        dwResult = GetExtendedTcpTable(pTCPInfo,
-                                       &dwSize,
-                                       FALSE,
-                                       AF_INET,
-                                       TCP_TABLE_OWNER_PID_LISTENER,
-                                       0);
-        if (dwResult != NO_ERROR)
-        {
-            hr = HRESULT_FROM_WIN32(dwResult);
-            goto Finished;
+            pTCPInfo = (MIB_TCPTABLE_OWNER_PID*)HeapAlloc(GetProcessHeap(), 0, dwSize);
+            if (pTCPInfo == NULL)
+            {
+                hr = E_OUTOFMEMORY;
+                goto Finished;
+            }
+
+            dwResult = GetExtendedTcpTable(pTCPInfo,
+                &dwSize,
+                FALSE,
+                AF_INET,
+                TCP_TABLE_OWNER_PID_LISTENER,
+                0);
+
+            if (dwResult != NO_ERROR && dwResult != ERROR_INSUFFICIENT_BUFFER)
```

Comment:
HeapFree at the end of the method if there is a failure? Also why not have the HeapFree at the end of the loop?

Reply:
it's safer as when retry happens we make sure the previous resource gets released first to avoid memory leak. if doing this at the end of loop, I have to do another condition check.